flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/logging/TestLoggerResource.java

```diff
@@ -43,52 +52,125 @@
+        this.loggerName = loggerName;
         this.level = level;
     }
 
     public List<String> getMessages() {
         return new ArrayList<>(loggingEvents);
     }
 
+    private static String generateRandomString() {
+        return UUID.randomUUID().toString().replace("-", "");
+    }
+
     @Override
     protected void before() throws Throwable {
         loggingEvents = new ConcurrentLinkedQueue<>();
 
+        final LoggerConfig previousLoggerConfig =
+                LOGGER_CONTEXT.getConfiguration().getLoggerConfig(loggerName);
+
+        final Level previousLevel = previousLoggerConfig.getLevel();
+        final Level userDefinedLevel = Level.getLevel(level.name());
+
+        // Set log level to last specific. This ensures that the parent still receives all log
+        // lines.
+        // WARN is more specific than INFO is more specific than DEBUG etc.
+        final Level newLevel =
+                userDefinedLevel.isMoreSpecificThan(previousLevel)
+                        ? previousLevel
+                        : userDefinedLevel;
+
+        // Filter log lines according to user requirements.
+        final Filter levelFilter =
+                ThresholdFilter.createFilter(
+                        userDefinedLevel, Filter.Result.ACCEPT, Filter.Result.DENY);
+
         Appender testAppender =
-                new AbstractAppender("test-appender", null, null, false) {
+                new AbstractAppender(
+                        "test-appender-" + generateRandomString(), levelFilter, null, false) {
                     @Override
                     public void append(LogEvent event) {
                         loggingEvents.add(event.getMessage().getFormattedMessage());
                     }
                 };
         testAppender.start();
 
-        AppenderRef appenderRef = AppenderRef.createAppenderRef(testAppender.getName(), null, null);
-        LoggerConfig logger =
+        LoggerConfig loggerConfig =
                 LoggerConfig.createLogger(
-                        false,
-                        Level.getLevel(level.name()),
-                        "test",
+                        true,
+                        newLevel,
+                        loggerName,
                         null,
-                        new AppenderRef[] {appenderRef},
+                        new AppenderRef[] {},
                         null,
                         LOGGER_CONTEXT.getConfiguration(),
                         null);
-        logger.addAppender(testAppender, null, null);
+        loggerConfig.addAppender(testAppender, null, null);
 
-        LOGGER_CONTEXT.getConfiguration().addLogger(loggerName, logger);
+        if (previousLoggerConfig.getName().equals(loggerName)) {
+            // remove the previous logger config for the duration of the test
+            backupLoggerConfig = previousLoggerConfig;
+            LOGGER_CONTEXT.getConfiguration().removeLogger(loggerName);
```

Comment:
What would happen if this line were removed? Will addLogger below fail, overwrite the existing logger, something else?

Reply:
In this case the old logger will not be replaced. At least this is what I see in my experiments. There is no clear documentation for this.